flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java

```diff
@@ -18,176 +18,1167 @@
-                                                                                "i",
-                                                                                DataTypes.INT()))),
-                                                        DataTypes.FIELD("s", DataTypes.STRING()))),
-                                Row.of(Row.of("12", true, null), "Hello"),
-                                DataTypes.of(
-                                        "ROW<r ROW<s STRING NOT NULL, b BOOLEAN, i INT>, s STRING>")),
-                TestSpec.forFunction(
-                                BuiltInFunctionDefinitions.CAST,
-                                "implicit between structured type and row")
-                        .onFieldsWithData(12, "Ingo")
-                        .withFunction(StructuredTypeConstructor.class)
-                        .withFunction(RowToFirstField.class)
-                        .testResult(
-                                call(
-                                        "RowToFirstField",
-                                        call("StructuredTypeConstructor", row($("f0"), $("f1")))),
-                                "RowToFirstField(StructuredTypeConstructor((f0, f1)))",
-                                12,
-                                DataTypes.INT()),
-                TestSpec.forFunction(
-                                BuiltInFunctionDefinitions.CAST,
-                                "explicit between structured type and row")
-                        .onFieldsWithData(12, "Ingo")
-                        .withFunction(StructuredTypeConstructor.class)
-                        .testTableApiResult(
-                                call("StructuredTypeConstructor", row($("f0"), $("f1")))
-                                        .cast(
-                                                DataTypes.ROW(
-                                                        DataTypes.BIGINT(), DataTypes.STRING())),
-                                Row.of(12L, "Ingo"),
-                                DataTypes.ROW(DataTypes.BIGINT(), DataTypes.STRING())));
-    }
+                CastTestSpecBuilder.testCast(CHAR(3), "To CHAR(3)")
+                        .from(CHAR(3), "foo")
+                        .resultsIn("foo")
+                        .from(CHAR(4), "foo")
+                        .resultsIn("foo ")
+                        .from(CHAR(4), "foo ")
+                        .resultsIn("foo ")
+                        .from(VARCHAR(3), "foo")
+                        .resultsIn("foo")
+                        .from(VARCHAR(5), "foo")
+                        .resultsIn("foo")
+                        .from(VARCHAR(5), "foo ")
+                        .resultsIn("foo ")
+                        // https://issues.apache.org/jira/browse/FLINK-24413 - Trim to precision
+                        // in this case down to 3 chars
+                        .from(STRING(), "abcdef")
+                        .resultsIn("abcdef")
```

Comment:
Perhaps can you comment this one like you did below for other wrong cases/results and write the correct result here?